### ISSUE TITLE
infiniband/rdma/verbs

### ISSUE THREAD

**Issue by eigood** [lizardfs/lizardfs#278]
Is there any interest in support for the above?  We've always wanted direct support in moosefs, but now we feel much better about lizardfs upstream, so might be willing to investigate this ourselves.

**Comment by psarna** [comment]
Hi Adam,

we did not consider supporting infiniband directly in near future. If you wish to investigate this issue yourself, you can certainly count on our help - but bear in mind, that we do not own any hardware to test infiniband yet.

**Comment by eigood** [comment]
Yeah, having hardware seems to be a requirement.  We only have first-gen infiniband, when can be gotten really cheap on ebay or elsewhere.

With IPoIB(TCP on infiniband) I can easily get 700MB/s, on the class of machines that we have.  This is using wget from apache on a sparse file, so no disk I/O.  Considering that we like to use nodes with 4 drive slots(1U), it seems like we don't actually need more recent infiniband with higher speeds.

**Comment by jwest75674** [comment]
@eigood , did you ever happen to make progress on Infiniband usage with Lizardfs? Alternatively, how did your usage end up settling in?

**Comment by eigood** [comment]
The usage pattern we saw is a combination of factors.  Most were what I would consider kernel bugs.

1: Let server run for a long time.  Have memory fragmentation occur(as shown via /proc/buddyinfo)
2: lizardfs FUSE requests a data block from the remote; it sends it as a 64k request.
3: local infiniband driver gets notified of an incoming 64k request, tries to allocate that much memory, fails.
4: Asks for a retry at 32k.  Fails.  16k fails, etc, etc, down to 4k.  Gets the single page, transfers just 4k.
5: Loop back to 3, with only 4k of data transfered.

Added the following 2 lines to cron to fix the above, and maintain good performance:
==
19 *    * * *   root    echo 1 > /proc/sys/vm/drop_caches
20 *    * * *   root    echo 1 > /proc/sys/vm/compact_memory
==

We have not upgraded our infrastructure, and we are slowly transitioning to a full cloud presence.  We love lizardfs, and might use it as part of a cloud-agnostic distributed FS, but in the short term, we are working towards not doing any self-hosting.

We are still interested in direct infiniband support, if it can be made to happen with relative ease.

**Comment by eigood** [comment]
Our basic assumption was to assume that the 1U machine *would fail*.  Power code, network cable, SSD backplane(we once had a capacitor burst, and set the circuit board on fire!).  So our local installs placed the OS on a 4-drive raid10, with a small partition, and the rest of the data on each was *directly* exposed to lizardfs via separate mounts on in mfshdd.cfg.

There is no local redundancy at all on a chunkserver.

We then had 2 power drops into the cabinet, split the servers into an A/B setup, and added a custom goal called "dual_ps", where we had 3 chunk copies, one on A, one on B, and the third someplace else.
